asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts

```diff
@@ -0,0 +1,325 @@
+                // return the zero range start with the correct end
+
+                const data: SliceResults = await determineSlice(cloneDates, slicerId, false);
+
+                return {
+                    start: dateParams.start,
+                    end: data.end,
+                    count: data.count
+                };
+            }
+
+            // find difference in milliseconds and divide in half
+            const diff = splitTime(dateParams.start, dateParams.end, dateParams.limit);
+            const newEnd = moment(dateParams.start).add(diff, timeResolution);
+
+            // prevent recursive call if difference is one millisecond
+            if (diff <= 0) {
+                return { start: dateParams.start, end: dateParams.end, count };
+            }
+
+            // recurse to find smaller chunk
+            dateParams.end = newEnd;
+            events.emit('slicer:slice:recursion');
+            logger.trace(`slicer: ${slicerId} is recursing ${JSON.stringify(dateParams)}`);
+
+            return determineSlice(dateParams, slicerId, isExpandedSlice);
+        }
+
+        // interval is only passed in with once mode, it will expand slices to prevent
+        // counts of 0, if the limit is reached it will run once more for the correct count
+        // then it should return and not recurse further if there is still no data
+        if (!isLimitQuery && count === 0 && dateParams.interval) {
+            // increase the slice range to find documents
+            let makeLimitQuery = false;
+
+            const newEnd = moment(dateParams.end).add(intervalNum, intervalUnit);
+            if (newEnd.isSameOrAfter(dateParams.limit)) {
+                // set to limit
+                makeLimitQuery = true;
+                dateParams.end = dateParams.limit;
+            } else {
+                dateParams.end = newEnd;
+            }
+            events.emit('slicer:slice:range_expansion');
+            return determineSlice(dateParams, slicerId, true, makeLimitQuery);
+        }
+
+        return { start: dateParams.start, end: dateParams.end, count };
+    }
+
+    async function getIdData(slicerFn: any) {
+        const list: any[] = [];
+        return new Promise(((resolve, reject) => {
+            const slicer = slicerFn;
+            function iterate() {
+                Promise.resolve(slicer())
+                    .then((data) => {
+                        if (data) {
+                            list.push(cloneDeep(data));
+                            return iterate();
+                        }
+                        return resolve(list);
+                    })
+                    .catch((err) => {
+                        // retries happen at the idSlicer level
+                        const errMessage = parseError(err);
+                        logger.error('error trying to subslice by key on getIdData:', errMessage);
+                        reject(errMessage);
```

Comment:
Don't throw with a string and use TSError
```suggestion
                        reject(new TSError(err, {
                            reason: 'error trying to subslice by key'
                        }));
```